helix-term/src/commands/typed.rs

```diff
@@ -18,7 +18,36 @@ pub struct TypableCommand {
     pub doc: &'static str,
     // params, flags, helper, completer
     pub fun: fn(&mut compositor::Context, &[Cow<str>], PromptEvent) -> anyhow::Result<()>,
-    pub completer: Option<Completer>,
+    /// What completion methods, if any, does this command have?
+    pub signature: CommandSignature,
+}
+
+impl TypableCommand {
+    fn completer_for_argument_number(&self, n: usize) -> Option<&Completer> {
+        self.signature
+            .positional_args
+            .get(n)
+            .unwrap_or(&self.signature.var_args)
+            .as_ref()
+    }
+}
+
+#[derive(Clone)]
+pub struct CommandSignature {
+    // Arguments with specific completion methods based on their position.
+    positional_args: &'static [Option<Completer>],
+
+    // All remaining arguments will use this completion method, if set.
+    var_args: Option<Completer>,
+}
+
+impl CommandSignature {
+    const fn none() -> Self {
+        Self {
+            positional_args: &[],
+            var_args: None,
+        }
+    }
```

Comment:
I think adding a couple more utility constructs will go a long way towards keeping the command list short and readable.

```suggestion
    const fn none() -> Self {
        Self {
            positional_args: &[],
            var_args: completers::none,
        }
    }
    const fn single(completer: &'static Completer) -> Self {
        Self {
            positional_args: slice::from_ref(completer),
            var_args: None,
        }
    }
    const fn repeat(completer: Completer) -> Self {
        Self {
            positional_args: &[],
            var_args: completer,
        }
    }
```

Reply:
You know it's funny I actually wrote basically the same thing first and wasn't sure if people would be comfortable with the layers of indirection. I agree and will put it back! :-)